pkgs/dart_mcp/example/file_system_server.dart

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io' as io;
+
+import 'package:async/async.dart';
+import 'package:dart_mcp/server.dart';
+import 'package:path/path.dart' as p;
+import 'package:stream_channel/stream_channel.dart';
+
+void main() {
+  SimpleFileSystemServer.fromStreamChannel(
+    StreamChannel.withCloseGuarantee(io.stdin, io.stdout)
+        .transform(StreamChannelTransformer.fromCodec(utf8))
+        .transformStream(const LineSplitter())
+        .transformSink(
+          StreamSinkTransformer.fromHandlers(
+            handleData: (data, sink) {
+              sink.add('$data\n');
+            },
+          ),
+        ),
+  );
+}
+
+/// An basic file server implementation.
+///
+/// Only allows reading/writing to the tracked [roots], and only by utf8 string.
+final class SimpleFileSystemServer extends MCPServer
+    with LoggingSupport, RootsTrackingSupport, ToolsSupport {
+  SimpleFileSystemServer.fromStreamChannel(super.channel)
+    : super.fromStreamChannel(
+        implementation: ServerImplementation(
+          name: 'file system',
+          version: '0.0.1',
+        ),
+      );
+
+  @override
+  FutureOr<InitializeResult> initialize(InitializeRequest request) {
+    registerTool(readFileTool, _readFile);
+    registerTool(writeFileTool, _writeFile);
+    registerTool(deleteFileTool, _deleteFile);
```

Comment:
Can we please add a tool for `ls`ing a directory within the allowed path?

Reply:
done